pkg/networkmanager/firewall.jsx

```diff
@@ -192,6 +208,165 @@ class AddServicesBody extends React.Component {
+     *      - description => _may be not present_ (Web Server)
+     */
+    parseServices(content) {
+        if (!content) {
+            console.warn("Couldn't read /etc/services");
+            return [ ];
+        }
+
+        var ret = {};
+        content.split('\n').forEach(line => {
+            if (!line || line.startsWith("#"))
+                return;
+            let m = line.match(/^(\S+)\s+(\d+)\/(\S+).*?(#(.*))?$/);
+            let new_port = { name: m[1], port: m[2], type: [m[3]] };
+            if (m.length > 5 && m[5])
+                new_port['description'] = m[5].trim();
+            if (ret[m[1]])
+                ret[m[1]]['type'].push(new_port['type'][0]);
+            else
+                ret[m[1]] = new_port;
+            if (ret[m[2]])
+                ret[m[2]]['type'].push(new_port['type'][0]);
+            else
+                ret[m[2]] = new_port;
+        });
+        return ret;
+    }
+
+    setName(event) {
+        this.setState({
+            custom_name: event.target.value,
+            generate_name: false,
+        });
+    }
+
+    getName(port) {
+        let known = this.state.avail_services[port];
+        if (known)
+            return known['name'];
+        else
+            return port;
+    }
+
+    generateName(all_ports) {
+        var name = "";
+        if (all_ports.length === 1) {
+            let known = this.state.avail_services[all_ports[0]];
+            if (known)
+                name = known['description'] || known['name'];
+            else
+                name = all_ports[0];
+        } else if (all_ports.length > 1) {
+            name = all_ports.slice(0, 3).map(this.getName)
+                    .join(", ");
+            if (all_ports.length > 3)
+                name += "...";
+        }
+        return name;
+    }
+
+    getPortNumber(port, type, avail) {
+        if (!avail) {
+            let num_p = Number(port);
+            if (isNaN(num_p))
+                return [0, _("Unknown service name")];
+            else if (num_p <= 0 || num_p > 65535)
+                return [0, _("Not valid port number")];
+            else
+                return [port, ""];
+        } else if (avail['type'].indexOf(type) < 0)
+            return [0, _("Port number and type do not match")];
+        else {
+            return [avail.port, ""];
+        }
+    }
+
+    validate(event) {
+        let ports = event.target.value.split(',');
+        let error = "";
+        let targets = ['tcp', 'custom_tcp_ports', 'tcp_error', 'custom_tcp_value'];
+        if (event.target.id === "udp-ports")
+            targets = ['udp', 'custom_udp_ports', 'udp_error', 'custom_udp_value'];
+
+        let new_ports = [];
+        let self = this;
+        ports.forEach(function(port) {
+            port = port.trim();
+            if (!port)
+                return;
+            let p1, p2;
+            if (port.indexOf("-") > -1) {
+                [p1, p2] = port.split("-");
```

Comment:
This does not currently handle the case of multiple `-`; e. g.  if you enter `10-50-500`, the `-500` just gets ignored. There should be an error message  instead.